Added documentation for exporting saved Metric Explorer charts with the XDMoD API (#1907)
ubccr/xdmod: docs/howto-api-image-export.md

The script will export your saved metric explorer charts to the current working directory.

```python
#!/usr/bin/env python3
import os
import requests
import json
import urllib
from dotenv import load_dotenv

load_dotenv()

username = os.getenv('XDMOD_USERNAME')
password = os.getenv('XDMOD_PASSWORD')
site_address = ""
image_format = "svg"

session = requests.Session()

auth_response = session.post(f'{site_address}/rest/auth/login', auth=(username, password))

if auth_response.status_code != 200:
    print('Authentication failed. Check provided credentials and check if you have a local XDMoD account')
    quit()

auth_response = auth_response.json()

header = {
  'Token': auth_response['results']['token'],
  'Authorization': auth_response['results']['token'],
  'Content-Type': 'application/x-www-form-urlencoded'
}

saved_charts = session.get(f'{site_address}/rest/v1/metrics/explorer/queries', headers=header, cookies=session.cookies)
saved_charts_data = saved_charts.json()

for idx, chart in enumerate(saved_charts_data['data']):
    if 'config' in chart:
        chart_json = json.loads(chart['config'])
        for attribute in chart_json:

>             if (isinstance(chart_json[attribute], dict)):

Reviewer: I suggest the `chart_json[attribute]` is stored in a variable for clarity/readability because of how often it is referenced here.

You can use the XDMoD API to image export your saved metric explorer charts. A local XDMoD account is **required** to authenticate through the API.

The following Python script will export your saved metric explorer charts. The `dotenv` library is recommended when authenticating through XDMoD API. You can install the `dotenv` library using:

`$ pip install python-dotenv`

Before running the script,

1. Create a `.env` file with your local XDMoD account credentials in the same directory as the script
1. Update `site_address` within the script with site address associated with your XDMoD instance.
1. Confirm the `image_format` within the script.

The script will export your saved metric explorer charts to the current working directory.

```python
#!/usr/bin/env python3
import os
import requests
import json
import urllib
from dotenv import load_dotenv

load_dotenv()

username = os.getenv('XDMOD_USERNAME')
password = os.getenv('XDMOD_PASSWORD')

> site_address = ""

Reviewer: Super minor, but I would suggest making this clearer that the `site_address` should be filled in. Either a note in the the top comment or `site_address = "<XDMOD_URL_HERE>"`. I assume most people will copy-paste this script without reading it and then encounter an issue when this isn't populated.

1. Confirm the `image_format` within the script.

The script will export your saved metric explorer charts to the current working directory. **\*Note:** Replace `<XDMOD_URL_HERE>` with your site address.

```python
#!/usr/bin/env python3
import os
import requests
import json
import urllib
from dotenv import load_dotenv

load_dotenv()

username = os.getenv('XDMOD_USERNAME')
password = os.getenv('XDMOD_PASSWORD')
site_address = "<XDMOD_URL_HERE>"
image_format = "svg"

session = requests.Session()

auth_response = session.post(f'{site_address}/rest/auth/login', auth=(username, password))

if auth_response.status_code != 200:
    print('Authentication failed. Check provided credentials and check if you have a local XDMoD account')
    quit()

auth_response = auth_response.json()

header = {
  'Token': auth_response['results']['token'],
  'Authorization': auth_response['results']['token'],
  'Content-Type': 'application/x-www-form-urlencoded'
}

saved_charts = session.get(f'{site_address}/rest/v1/metrics/explorer/queries', headers=header, cookies=session.cookies)
saved_charts_data = saved_charts.json()

> for idx, chart in enumerate(saved_charts_data['data']):

Reviewer 2: I would suggest adding an option to export a single chart with a given name rather than all the charts.

> The following Python script will export your saved Metric Explorer charts. An XDMoD username and password is required to run the script.

Reviewer 2: ```suggestion
The following Python script can be used to export your saved Metric Explorer charts to image files using the XDMoD REST API. An XDMoD username and password are required to run the script.
```

Before running the script,

> 1. Install the required `python-dotenv` and `requests` dependencies.

Reviewer 2: ```suggestion
1. Install the required `python-dotenv` and `requests` Python dependencies (e.g., using `pip`).
```

1. Create a `.env` file in the same directory as the script that contains the following contents, replacing `<username>` with your XDMoD username and `<password>` with your XDMoD password — make sure to secure this file with read-only access.
    ```
    XDMOD_USERNAME=<username>
    XDMOD_PASSWORD=<password>
    ```

> 1. Update the value of `site_address` within the script with the URL associated with your XDMoD portal.
> 1. Update the value of `export_path` within the script with the desired path to export the images.
> 1. Confirm the `image_format` within the script.
> 
> The default image format is `svg`, but `png` and `pdf` formats are also supported. Refer to the XDMoD [Metric Explorer Tab Controller API](rest.html#tag/Metric-Explorer/paths/~1controllers~1metric_explorer.php/post) `get_data` operation for more information on the request body schema.

Reviewer 2: ```suggestion
1. Update the value of `site_address` at the top of the script with the URL associated with your XDMoD portal.
1. Update the value of `export_path` at the top of the script with the desired directory path where the images will be written.
1. Confirm the desired values for `image_format`, `width`, and `height` at the top of the script. The default image format is `svg`, but `png` and `pdf` formats are also supported.

By default, the script will download all of your saved Metric Explorer charts. You can have it instead download a single chart by providing the `-n` or `--name` option followed by the name of the saved chart.

Refer to the XDMoD [Metric Explorer Tab Controller REST API](rest.html#tag/Metric-Explorer/paths/~1controllers~1metric_explorer.php/post) `get_data` operation for more information on the REST request body schema.
```

```python
#!/usr/bin/env python3
import os
import requests
import json
import urllib
import argparse

> from dotenv import load_dotenv

Reviewer 2: ```suggestion
from dotenv import load_dotenv
import sys
```

site_address = ''
export_path = ''
image_format = 'svg'
image_width = 916
image_height = 484

> 

Reviewer 2: ```suggestion

if site_address == '':
    print('Please edit the script and specify a site_address.', file=sys.stderr)
    sys.exit(1)

```

load_dotenv()
username = os.getenv('XDMOD_USERNAME')
password = os.getenv('XDMOD_PASSWORD')

parser = argparse.ArgumentParser(description='Export XDMoD saved Metric Explorer charts with the REST API.')

> parser.add_argument('-n', '--name',type=str, default='', help='Specify the chart name of a saved chart to export.')

Reviewer 2: ```suggestion
parser.add_argument('-n', '--name', help='Specify the chart name of a saved chart to export.')
```
The type will already be string, and `None` will be the default. For simplicity, I'd stick with these.

args = parser.parse_args()

session = requests.Session()

auth_response = session.post(f'{site_address}/rest/auth/login', auth=(username, password))

if auth_response.status_code != 200:

>     print('Authentication failed. Check provided credentials and check if you have a local XDMoD account')
>     quit()

Reviewer 2: ```suggestion
    print('Authentication failed. Check provided credentials.', file=sys.stderr)
    sys.exit(1)
```

auth_response = auth_response.json()

header = {
    'Token': auth_response['results']['token'],
    'Authorization': auth_response['results']['token'],
    'Content-Type': 'application/x-www-form-urlencoded'
}

saved_charts = session.get(f'{site_address}/rest/v1/metrics/explorer/queries', headers=header)
saved_charts_data = saved_charts.json()

> if args.name != '' and not any(chart_obj['name'] == args.name for chart_obj in saved_charts_data['data']):

Reviewer 2: ```suggestion
if args.name is not None and not any(chart_obj['name'] == args.name for chart_obj in saved_charts_data['data']):
```

>     print('Specified chart not found.')
>     exit()

Reviewer 2: ```suggestion
    print('Specified chart not found.', file=sys.stderr)
    sys.exit(1)
```

> for idx, chart in enumerate(saved_charts_data['data']):

Reviewer 2: ```suggestion
for idx, chart in enumerate(saved_charts_data['data']):
    if args.name is not None and args.name != chart['name']:
        continue
```
and then you don't need to wrap the writing in an if/else or needlessly fetch each chart that isn't going to be written.

            if (isinstance(chart_parameter, dict)):
                if 'data' in attribute:
                    encoded_str = urllib.parse.quote_plus(str(chart_parameter['data']))
                else:
                    encoded_str = urllib.parse.quote_plus(str(chart_parameter))
                encoded_str = encoded_str.replace('%27','%22').replace('False', 'false').replace('True', 'true').replace('None', 'null')
                chart_json[attribute] = encoded_str
            if chart_parameter in (True, False, None):
                chart_json[attribute] = str(chart_parameter).replace('False', 'false').replace('True', 'true').replace('None', 'null')

        chart_json['operation'] = "get_data"
        chart_json['controller_module'] = "metric_explorer"
        chart_json['show_title'] = "y"
        chart_json['format'] = image_format
        chart_json['width'] = image_width
        chart_json['height'] = image_height

        chart_response = session.post(f'{site_address}/controllers/metric_explorer.php', data=chart_json, headers=header)
        chart_name = f"{chart['name']}.{image_format}" if ('name' in chart) else f"xdmod_API_export_{idx}.{image_format}"

>         if args.name != '' and args.name == chart['name']:
>             with open(export_path + chart_name, "wb") as f:
>                 f.write(chart_response.content)
>                 exit()
>         else:
>             with open(export_path + chart_name, "wb") as f:
>                 f.write(chart_response.content)

Reviewer 2: ```suggestion
        with open(export_path + chart_name, "wb") as f:
            f.write(chart_response.content)
```
See suggestion above with `continue`.
